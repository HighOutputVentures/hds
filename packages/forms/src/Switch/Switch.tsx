import {
  Box,
  FormControl,
  FormHelperText,
  FormLabel,
  MultiStyleConfig,
  Switch as ChakraSwitch,
} from "@chakra-ui/react";
import * as React from "react";

type Size = "sm" | "md";

type ColorScheme = "dark" | "light";

export type SwitchProps = {
  size?: Size;
  isChecked?: boolean;
  isDisabled?: boolean;
  onCheck?(value: boolean): void;
  hint?: string;
  label?: string;
  colorScheme?: ColorScheme;
};

export default React.forwardRef<HTMLInputElement, SwitchProps>(function Switch(
  props,
  ref,
) {
  const {
    size = "md",
    isChecked,
    isDisabled,
    onCheck,
    hint,
    label,
    colorScheme = "light",
  } = props;

  const uniqid = React.useId();
  const styleConfig = useSwitchStyleConfig({
    size,
    colorScheme,
  });

  return (
    <FormControl
      variant="unstyled"
      sx={{
        display: "flex",
        ...(size === "sm" && { gap: "8px" }),
        ...(size === "md" && { gap: "12px" }),
      }}
      data-testid="hds.switch.form-group"
    >
      <ChakraSwitch
        ref={ref}
        id={uniqid}
        size={size}
        isChecked={isChecked}
        isDisabled={isDisabled}
        onChange={(e) => {
          onCheck?.(e.target.checked);
        }}
        styleConfig={styleConfig}
        data-testid="hds.switch"
      />

      <Box>
        {!!label && (
          <FormLabel
            htmlFor={uniqid}
            sx={{
              color: "Gray.700",
              fontWeight: "medium",

              ...(size === "sm" && {
                margin: "0px",
                fontSize: "14px",
                lineHeight: "20px",
              }),
              ...(size === "md" && {
                margin: "0px",
                fontSize: "16px",
                lineHeight: "24px",
              }),
            }}
            data-testid="hds.switch.label"
          >
            {label}
          </FormLabel>
        )}

        {!!hint && (
          <FormHelperText
            sx={{
              color: "Gray.500",
              ...(size === "sm" && {
                fontSize: "14px",
                lineHeight: "20px",
                marginTop: "0px",
              }),
              ...(size === "md" && {
                fontSize: "16px",
                lineHeight: "24px",
                marginTop: "2px",
              }),
            }}
            data-testid="hds.switch.hint"
          >
            {hint}
          </FormHelperText>
        )}
      </Box>
    </FormControl>
  );
});

type UseSwitchStyleConfigArg = {
  size: Size;
  colorScheme: ColorScheme;
};

function useSwitchStyleConfig(arg: UseSwitchStyleConfigArg) {
  const styleConfig = React.useMemo<MultiStyleConfig>(
    function getStyleConfig() {
      return {
        parts: ["container", "thumb", "track"],
        sizes: {},
        variants: {},
        defaultProps: {},
        baseStyle: {
          container: {},
          track: {
            padding: "2px",
            rounded: "full",

            ...(arg.size === "sm" && { width: "32px", height: "16px" }),
            ...(arg.size === "md" && { width: "40px", height: "20px" }),

            ...(arg.colorScheme === "light" && {
              bgColor: "neutrals.200",
              _hover: {
                bgColor: "neutrals.100",
              },
              _focus: {
                bgColor: "Gray.50",
                boxShadow: "0px 0px 0px 4px #F4EBFF",
              },
              _checked: {
                bgColor: "brand.primary.700",
                _hover: {
                  bgColor: "brand.primary.900",
                },
              },
              "&[data-disabled]": {
                bgColor: "#F0F0F0!important",
                opacity: 1,
              },
            }),

            ...(arg.colorScheme === "dark" && {
              bgColor: "brand.primary.50",
              _hover: {
                bgColor: "brand.primary.100",
              },
              _focus: {
                boxShadow: "0px 0px 0px 4px #F4EBFF",
              },
              _checked: {
                bgColor: "brand.primary.200",
                _hover: {
                  bgColor: "brand.primary.300",
                },
              },
              "&[data-disabled]": {
                bgColor: "#F0F0F0!important",
                opacity: 1,
              },
            }),
            transition:
              "background-color 300ms ease-in-out, box-shadow 300ms ease-in-out",
          },
          thumb: {
            ...(arg.size === "sm" && {
              width: "16px",
              height: "16px",
            }),

            ...(arg.size === "md" && {
              width: "20px",
              height: "20px",
            }),

            rounded: "full",
            padding: "0px",
            bgColor: "white!important",
            transition: "all 300ms ease-in-out",
            boxShadow:
              "0px 1px 3px rgba(16, 24, 40, 0.1), 0px 1px 2px rgba(16, 24, 40, 0.06)",
            _checked: {
              transform: "translateX(100%)",
            },
          },
        },
      };
    },
    [arg],
  );

  return styleConfig;
}
