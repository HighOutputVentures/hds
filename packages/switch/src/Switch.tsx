import {
  FormControl,
  FormHelperText,
  FormLabel,
  Switch,
  SystemStyleObject,
  useMultiStyleConfig,
} from "@chakra-ui/react";
import * as React from "react";

type SwitchValue = "on" | "off";

type SwitchSize = "sm" | "md";

type ResponsiveSize<T> = Record<"base" | "sm" | "md" | "lg" | "xl", T>;

export type SwitchBaseProps = {
  size?: SwitchSize | ResponsiveSize<SwitchSize>;
  value?: SwitchValue;
  onToggle?(value: SwitchValue): void;
  label?: string;
  helperText?: string;
};

export type SwitchProps = SwitchBaseProps & SystemStyleObject;

export default React.forwardRef<HTMLInputElement, SwitchProps>(function HdsSwitch(props, ref) {
  const { size, value, onToggle, label, helperText, ...others } = Object.assign(
    defaultProps,
    props,
  );

  const styles = useMultiStyleConfig("Switch", { size, variant: "hds" });
  const uniqid = React.useId();

  return (
    <FormControl sx={{ ...others, ...styles.control }}>
      <Switch
        ref={ref}
        id={uniqid}
        isChecked={value === "on"}
        onChange={(e) => {
          onToggle(e.target.checked ? "on" : "off");
        }}
      />

      {label && (
        <FormLabel htmlFor={uniqid} sx={styles.label}>
          {label}
        </FormLabel>
      )}
      {helperText && <FormHelperText sx={styles.helperText}>{helperText}</FormHelperText>}
    </FormControl>
  );
});

const defaultProps = {
  size: "md",
  value: "off",
  onToggle() {},
};
