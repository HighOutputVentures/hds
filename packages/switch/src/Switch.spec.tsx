import "@testing-library/jest-dom";
import { fireEvent, render, waitFor } from "@testing-library/react";
import * as React from "react";
import Switch, { SwitchProps } from "./Switch";

describe("Switch", () => {
  const renderTestComponent = (props: SwitchProps = {}) => {
    return render(<Switch {...props} />);
  };

  it("Should render FormControl", () => {
    const { queryByRole } = renderTestComponent();
    expect(queryByRole("group")).toBeDefined();
  });

  it("Should render switch", () => {
    const { queryByRole } = renderTestComponent();
    expect(queryByRole("checkbox")).toBeDefined();
  });

  it("Should be able to set checked", () => {
    const { getByRole } = renderTestComponent({ checked: true });
    expect(getByRole("checkbox")).toHaveAttribute("checked");
  });

  it("Should be able to set label", () => {
    const { queryByText } = renderTestComponent({ label: "Checkbox label" });
    expect(queryByText("Checkbox label")).toBeDefined();
  });

  it("Should be able to set helper text", () => {
    const { queryByText } = renderTestComponent({ helperText: "Checkbox helper text" });
    expect(queryByText("Checkbox helper text")).toBeDefined();
  });

  it("Should call callback fn when check is toggled", async () => {
    const onCheck = jest.fn();
    const { getByRole } = renderTestComponent({ onCheck });

    fireEvent.click(getByRole("checkbox"));

    await waitFor(() => {
      expect(onCheck).toHaveBeenCalledWith(expect.any(Boolean));
    });
  });
});
