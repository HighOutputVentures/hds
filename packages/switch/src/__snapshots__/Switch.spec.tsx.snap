// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch Should match snapshot 1`] = `
<div
  className="chakra-form-control css-0"
  role="group"
>
  <label
    className="chakra-switch css-16pgy8f"
    data-checked=""
    onClick={[Function]}
  >
    <input
      aria-disabled={false}
      aria-invalid={false}
      checked={true}
      className="chakra-switch__input"
      disabled={false}
      id=":r0:"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      readOnly={false}
      required={false}
      style={
        Object {
          "border": "0",
          "clip": "rect(0, 0, 0, 0)",
          "height": "1px",
          "margin": "-1px",
          "overflow": "hidden",
          "padding": "0",
          "position": "absolute",
          "whiteSpace": "nowrap",
          "width": "1px",
        }
      }
      type="checkbox"
    />
    <span
      aria-hidden={true}
      className="chakra-switch__track css-14qxnv8"
      data-checked=""
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
    >
      <span
        className="chakra-switch__thumb css-0"
        data-checked=""
      />
    </span>
  </label>
  <div
    className="css-0"
  >
    <label
      className="chakra-form__label css-uvho5g"
      htmlFor=":r0:"
      id="field-:r1:-label"
    >
      Checkbox label
    </label>
    <div
      className="chakra-form__helper-text css-0"
      id="field-:r1:-helptext"
    >
      Checkbox helper text
    </div>
  </div>
</div>
`;
